Added APIs for bootstrapping

crockeea opened the pull request:
*Issue #, if available:*

*Description of changes:*
This PR adds support for bootstrapping. It introduces the new types `Bootstrapper`, `BootstrappingKey` and `BootstrappingParameters`, plus associated APIs. Added a new example that mirrors the Lattigo example for bootstrapping.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

bryce-shang commented on src/latticpp/ckks/params.h:
```diff
@@ -29,13 +29,21 @@ namespace latticpp {
         // PN16QP1654pq is the index in DefaultParams for logQP = 1654 (post quantum)
-        PN16QP1654pq
+        PN16QP1654pq,
+        // Bootstrapping parameters index 0
+        BootstrapParams0,
+        // Bootstrapping parameters index 1
+        BootstrapParams1,
+        // Bootstrapping parameters index 2
+        BootstrapParams2,
+        // Bootstrapping parameters index 3
+        BootstrapParams3
```
Nit: Why there are only 4 `BootstrapParams`? Will we add BootstrapParams4?

crockeea replied:
These enums match corresponding default parameter sets in Lattigo, so these would only change if I change the version of Lattigo on which we depend.

bryce-shang commented on src/latticpp/ckks/bootstrap_params.h:
```diff
@@ -0,0 +1,25 @@
+namespace latticpp {
+
+    enum NamedBootstrappingParams {
+        // Bootstrapping parameters index 0
+        BootstrapParams_Set2,
+        // Bootstrapping parameters index 1
+        BootstrapParams_Set5,
+        // Bootstrapping parameters index 2
+        BootstrapParams_Set7,
+        // Bootstrapping parameters index 3
+        BootstrapParams_Set4
```
Why the set number and the index num are different?

crockeea replied:
The names (`Set*`) are from Lattigo, and are ordered in the same way. Presumably, these sets were given in their paper listed in some order, and they only included a subset of the published parameters in their implementation. As to why they reordered them in their code, I couldn't say. But for our enum to match their enum, we have to use the same order (and names, so that the parameter sets don't lose their semantic meaning relative to the paper which describes what is in each parameter set).

crockeea commented:
Prior to merging, I have to figure out the CI failure and fix a runtime discrepancy with the bootstrapping example. The CI failure is likely due to out-of-memory; I didn't run into a problem on my dev machine.

crockeea commented:
> Prior to merging, I have to figure out the CI failure and fix a runtime discrepancy with the bootstrapping example. The CI failure is likely due to out-of-memory; I didn't run into a problem on my dev machine.

Fixed CI problem by increasing the CI instance size. Turns out bootstrapping keys are quite large (~25+GB).

The precision error was caused by using "dense" HE keys with "sparse-key" bootstrapping parameters. To keep the bootstrapping example as close as possible to the Lattigo example, I added a new API for generating sparse keys; the example now works just like the Lattigo example.
